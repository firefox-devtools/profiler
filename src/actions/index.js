/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
