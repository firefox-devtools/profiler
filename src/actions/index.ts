/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import * as app from './app';
import * as icons from './icons';
import * as l10n from './l10n';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  l10n,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
