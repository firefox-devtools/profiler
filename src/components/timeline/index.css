/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.timelineOverflowEdgeIndicatorScrollbox {
  position: absolute;

  /* if the platform scrollbar ends up being larger than --vertical-scrollbar-reserved-width (which is unexpected), make sure there is no horizontal scrollbar */
  overflow: hidden auto;
  width: 100%;
  height: 100%;
  padding-left: var(--thread-label-column-width);
  margin-right: calc(var(--vertical-scrollbar-reserved-width) * -1);
  margin-left: calc(var(--thread-label-column-width) * -1);
  background: linear-gradient(
    to left,
    white,
    #eee calc(var(--vertical-scrollbar-reserved-width) - 1px),
    var(--grey-30) 0,
    var(--grey-30) var(--vertical-scrollbar-reserved-width),
    transparent 0
  );
}

.timelineThreadList {
  width: calc(100vw - var(--vertical-scrollbar-reserved-width));
  padding: 0;
  margin: 0 0 0 calc(var(--thread-label-column-width) * -1);
  background-color: var(--grey-20);
  box-shadow: inset 0 1px var(--grey-30);
  list-style: none;
}

.timelineSettingsHiddenTracks {
  /* Make sure and override all rules that have to do with button defaults. */
  position: relative;
  width: var(--thread-label-column-width);

  /* 8px is intentional to align this text with the text above. */
  padding: 0 20px 0 8px;
  border: 0;

  /* Separator line between tracks menu button and timeline ruler */
  border-right: 1px solid var(--grey-30);
  background: none;
  color: inherit;
  font: inherit;
  text-align: left;
}

.timelineSettingsHiddenTracks:hover {
  background-color: rgb(0 0 0 / 0.1);
}

.timelineSettingsHiddenTracks:hover:active {
  background-color: rgb(0 0 0 / 0.2);
}

/* This is the dropdown arrow on the left of the button. */
.timelineSettingsHiddenTracks::before {
  position: absolute;
  top: 2px;
  right: 2px;
  border-top: 6px solid;
  border-right: 4px solid transparent;
  border-bottom: 0 solid transparent;
  border-left: 4px solid transparent;
  margin-top: 5px;
  margin-right: 5px;
  margin-left: 5px;
  color: var(--grey-60);
  content: '';

  /* Opt-out of forced colors so arrow is visible in High Contrast Mode */
  forced-color-adjust: none;
}

.timelineSettingsHiddenTracksNumber {
  font-weight: bold;
}

.timelineHeader {
  /* This should be on top of the overflow edge indicator that wraps the thread
   list, so border bottom will always be visible. */
  z-index: 1;
  display: flex;
  height: 20px;
  flex-direction: row;

  /* Timeline header content should end before the scrollbar but also border
   * bottom should extend to the whole timeline header. Because of that, we are
   * creating a padding and a minus margin with the scrollbar width to make
   * this work. */
  padding-right: var(--vertical-scrollbar-reserved-width);
  border-bottom: 1px solid var(--grey-30);
  margin-right: calc(var(--vertical-scrollbar-reserved-width) * -1);
  margin-left: calc(var(--thread-label-column-width) * -1);

  /* This div needs to be apaque so that it can properly hide the hidden tracks
   * container border below */
  background: white;
}

.tracksContainer {
  /* This should be below the timeline header. */
  z-index: 0;

  /* Moving the timeline 1 pixel below the timeline header. This way, the first
   * visible track's top border will be hidden as expected. */
  margin-top: -1px;
  margin-bottom: -1px;
}

@media (forced-colors: active) {
  .timelineSettingsHiddenTracks {
    border: 1px solid currentcolor;
    background-color: ButtonFace;
    color: ButtonText;
  }

  .timelineSettingsHiddenTracks:hover {
    background-color: SelectedItemText;
    color: SelectedItem;
  }

  .timelineSettingsHiddenTracks:hover:active {
    border-color: ButtonText;
  }

  .timelineSettingsHiddenTracks::before {
    color: currentcolor;
  }
}
