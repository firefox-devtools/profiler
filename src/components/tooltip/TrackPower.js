/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import { Localized } from '@fluent/react';
import memoize from 'memoize-one';

import { averageIntensity } from '@tgwf/co2';

import explicitConnect from 'firefox-profiler/utils/connect';
import { formatNumber } from 'firefox-profiler/utils/format-numbers';
import {
  getCommittedRange,
  getPreviewSelection,
  getProfileInterval,
  getMeta,
} from 'firefox-profiler/selectors/profile';
import { getSampleIndexRangeForSelection } from 'firefox-profiler/profile-logic/profile-data';

import { TooltipDetails, TooltipDetail } from './TooltipDetails';

import type {
  Counter,
  Milliseconds,
  PreviewSelection,
  StartEndRange,
  ProfileMeta,
} from 'firefox-profiler/types';

import type { ConnectedProps } from 'firefox-profiler/utils/connect';

type OwnProps = {
  counter: Counter,
  counterSampleIndex: number,
};

type StateProps = {
  interval: Milliseconds,
  meta: ProfileMeta,
  committedRange: StartEndRange,
  previewSelection: PreviewSelection,
};

type Props = ConnectedProps<OwnProps, StateProps, {}>;

class TooltipTrackPowerImpl extends React.PureComponent<Props> {
  // This compute the sum of the power in the range. This returns a value in Wh.
  _computePowerSumForRange(start: Milliseconds, end: Milliseconds): number {
    const { counter } = this.props;
    const samples = counter.samples;
    const [beginIndex, endIndex] = getSampleIndexRangeForSelection(
      samples,
      start,
      end
    );

    let sum = 0;
    for (
      let counterSampleIndex = beginIndex;
      counterSampleIndex < endIndex;
      counterSampleIndex++
    ) {
      sum += samples.count[counterSampleIndex]; // picowatt-hour;
    }
    return sum * 1e-12;
  }

  _computeCO2eFromPower(power: number): number {
    // total energy Wh to kWh
    const energy = power / 1000;
    const intensity =
      this.props.meta.gramsOfCO2ePerKWh || averageIntensity.data.WORLD;
    return energy * intensity;
  }

  _computePowerSumForCommittedRange = memoize(
    ({ start, end }: StartEndRange): number =>
      this._computePowerSumForRange(start, end)
  );

  _formatPowerValue(
    power: number,
    l10nIdKiloUnit,
    l10nIdUnit,
    l10nIdMilliUnit,
    l10nIdMicroUnit
  ): Localized {
    let value, l10nId, carbonValue;
    const carbon = this._computeCO2eFromPower(power);
    if (power > 1000) {
      value = formatNumber(power / 1000, 3);
      carbonValue = formatNumber(carbon / 1000, 2);
      l10nId = l10nIdKiloUnit;
    } else if (power > 1) {
      value = formatNumber(power, 3);
      carbonValue = formatNumber(carbon, 3);
      l10nId = l10nIdUnit;
    } else if (power === 0) {
      value = 0;
      carbonValue = 0;
      l10nId = l10nIdUnit;
    } else if (power < 0.001 && l10nIdMicroUnit) {
      value = formatNumber(power * 1000000);
      // Note: even though the power value is expressed in ÂµWh, the carbon value
      // is still expressed in mg.
      carbonValue = formatNumber(carbon * 1000);
      l10nId = l10nIdMicroUnit;
    } else {
      value = formatNumber(power * 1000);
      carbonValue = formatNumber(carbon * 1000);
      l10nId = l10nIdMilliUnit;
    }

    return (
      <Localized
        id={l10nId}
        vars={{ value, carbonValue }}
        attrs={{ label: true }}
      >
        <TooltipDetail label="">{value}</TooltipDetail>
      </Localized>
    );
  }

  maybeRenderForPreviewSelection(
    previewSelection
  ): React.ChildrenArray<React.Element<typeof TooltipDetail> | null> | null {
    if (!previewSelection.hasSelection) {
      return null;
    }

    const { selectionStart, selectionEnd } = previewSelection;
    const selectionRange = selectionEnd - selectionStart;

    if (selectionRange === 0) {
      return null;
    }

    const powerSumForPreviewRange = this._computePowerSumForRange(
      selectionStart,
      selectionEnd
    );

    return (
      // $FlowExpectError our version of Flow doesn't understand Fragments very well.
      <>
        {this._formatPowerValue(
          powerSumForPreviewRange,
          'TrackPower--tooltip-energy-carbon-used-in-preview-kilowatthour',
          'TrackPower--tooltip-energy-carbon-used-in-preview-watthour',
          'TrackPower--tooltip-energy-carbon-used-in-preview-milliwatthour',
          'TrackPower--tooltip-energy-carbon-used-in-preview-microwatthour'
        )}
        {this._formatPowerValue(
          (1000 /* ms -> s */ * 3600 /* s -> h */ * powerSumForPreviewRange) /
            selectionRange,
          'TrackPower--tooltip-average-power-kilowatt',
          'TrackPower--tooltip-average-power-watt',
          'TrackPower--tooltip-average-power-milliwatt'
        )}
      </>
    );
  }

  render() {
    const {
      counter,
      counterSampleIndex,
      interval,
      committedRange,
      previewSelection,
    } = this.props;
    const samples = counter.samples;

    const powerUsageInPwh = samples.count[counterSampleIndex]; // picowatt-hour
    const sampleTimeDeltaInMs =
      counterSampleIndex === 0
        ? interval
        : samples.time[counterSampleIndex] -
          samples.time[counterSampleIndex - 1];
    const power =
      ((powerUsageInPwh * 1e-12) /* pWh->Wh */ / sampleTimeDeltaInMs) *
      1000 * // ms->s
      3600; // s->h

    return (
      <div className="timelineTrackPowerTooltip">
        <TooltipDetails>
          {this._formatPowerValue(
            power,
            'TrackPower--tooltip-power-kilowatt',
            'TrackPower--tooltip-power-watt',
            'TrackPower--tooltip-power-milliwatt'
          )}
          {this.maybeRenderForPreviewSelection(previewSelection)}
          {this._formatPowerValue(
            this._computePowerSumForCommittedRange(committedRange),
            'TrackPower--tooltip-energy-carbon-used-in-range-kilowatthour',
            'TrackPower--tooltip-energy-carbon-used-in-range-watthour',
            'TrackPower--tooltip-energy-carbon-used-in-range-milliwatthour',
            'TrackPower--tooltip-energy-carbon-used-in-range-microwatthour'
          )}
        </TooltipDetails>
      </div>
    );
  }
}

export const TooltipTrackPower = explicitConnect<OwnProps, StateProps, {}>({
  mapStateToProps: (state) => ({
    interval: getProfileInterval(state),
    meta: getMeta(state),
    committedRange: getCommittedRange(state),
    previewSelection: getPreviewSelection(state),
  }),
  component: TooltipTrackPowerImpl,
});
