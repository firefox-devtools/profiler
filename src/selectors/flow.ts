/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';

import {
  computeProfileFlowInfo,
  computeFlowTiming,
} from '../profile-logic/marker-data';
import { getThreadSelectors } from './per-thread';
import { getActiveFlows } from './url-state';
import type { ThreadSelectors } from './per-thread';
import {
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
} from './profile';

import type {
  Selector,
  State,
  MarkerIndex,
  Marker,
  ProfileFlowInfo,
  FlowTiming,
} from 'firefox-profiler/types';

function _arraysShallowEqual(arr1: any[], arr2: any[]): boolean {
  return arr1.length === arr2.length && arr1.every((val, i) => val === arr2[i]);
}

function _createSelectorForAllThreads<T>(
  f: (selectors: ThreadSelectors, state: State) => T
): Selector<T[]> {
  let previousOutputPerThread: T[] = [];
  return function recomputeSelectorForAllThreads(state: State): T[] {
    const threads = getThreads(state);
    let outputPerThread = threads.map((_thread, i) => {
      const threadSelectors = getThreadSelectors(i);
      return f(threadSelectors, state);
    });
    if (_arraysShallowEqual(outputPerThread, previousOutputPerThread)) {
      outputPerThread = previousOutputPerThread;
    }
    previousOutputPerThread = outputPerThread;
    return outputPerThread;
  };
}

export const getFullMarkerListPerThread: Selector<Marker[][]> =
  _createSelectorForAllThreads(({ getFullMarkerList }, state) =>
    getFullMarkerList(state)
  );

export const getMarkerChartLabelGetterPerThread: Selector<
  Array<(marker: MarkerIndex) => string>
> = _createSelectorForAllThreads(({ getMarkerChartLabelGetter }, state) =>
  getMarkerChartLabelGetter(state)
);

export const getProfileFlowInfo: Selector<ProfileFlowInfo> = createSelector(
  getFullMarkerListPerThread,
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
  computeProfileFlowInfo
);

export const getFlowTiming: Selector<FlowTiming> = createSelector(
  getProfileFlowInfo,
  getActiveFlows,
  computeFlowTiming
);
