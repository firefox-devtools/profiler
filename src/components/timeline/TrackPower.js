/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import explicitConnect from 'firefox-profiler/utils/connect';
import {
  getCommittedRange,
  getCounterSelectors,
} from 'firefox-profiler/selectors/profile';
import { updatePreviewSelection } from 'firefox-profiler/actions/profile-view';
import { TrackPowerGraph } from './TrackPowerGraph';
import {
  TRACK_POWER_HEIGHT,
  TRACK_POWER_LINE_WIDTH,
} from 'firefox-profiler/app-logic/constants';

import type {
  CounterIndex,
  ThreadIndex,
  Milliseconds,
} from 'firefox-profiler/types';

import type { ConnectedProps } from 'firefox-profiler/utils/connect';

import './TrackPower.css';

type OwnProps = {
  +counterIndex: CounterIndex,
};

type StateProps = {
  +threadIndex: ThreadIndex,
  +rangeStart: Milliseconds,
  +rangeEnd: Milliseconds,
};

type DispatchProps = {
  updatePreviewSelection: typeof updatePreviewSelection,
};

type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;

type State = {};

export class TrackPowerImpl extends React.PureComponent<Props, State> {
  render() {
    const { counterIndex } = this.props;
    return (
      <div
        className="timelineTrackPower"
        style={{
          height: TRACK_POWER_HEIGHT,
          '--graph-height': `${TRACK_POWER_HEIGHT}px`,
        }}
      >
        <TrackPowerGraph
          counterIndex={counterIndex}
          lineWidth={TRACK_POWER_LINE_WIDTH}
          graphHeight={TRACK_POWER_HEIGHT}
        />
      </div>
    );
  }
}

export const TrackPower = explicitConnect<OwnProps, StateProps, DispatchProps>({
  mapStateToProps: (state, ownProps) => {
    const { counterIndex } = ownProps;
    const counterSelectors = getCounterSelectors(counterIndex);
    const counter = counterSelectors.getCounter(state);
    const { start, end } = getCommittedRange(state);
    return {
      threadIndex: counter.mainThreadIndex,
      rangeStart: start,
      rangeEnd: end,
    };
  },
  mapDispatchToProps: { updatePreviewSelection },
  component: TrackPowerImpl,
});
