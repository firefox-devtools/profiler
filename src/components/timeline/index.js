/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import { FullTimeline } from 'firefox-profiler/components/timeline/FullTimeline';

type TimelineProps = {||};

export class Timeline extends PureComponent<TimelineProps> {
  // This may contain a function that's called whenever we want to remove the
  // "wheel" listener.
  _removeWheelListener: null | (() => mixed) = null;

  // This effectively disable the pinch-to-zoom as well as ctrl+mousewheel
  // gestures. Indeed in the timeline it is confusing. In the future we'll want
  // to couple this with the preview selection like in the Viewport HOC.
  preventPinchToZoom(e: WheelEvent) {
    if (e.ctrlKey) {
      e.preventDefault();
    }
  }

  // This will be registered as a ref property to the DOM element displaying the
  // tracks. We use this solution of registering the wheel event in this ref
  // listener because:
  // * we can't use React's event handling, because it doesn't allow us to use
  //   the "passive: false" way of registering the event handler. But we need this
  //   if we want to be able to prevent the default action of page zooming.
  // * we want to be sure to register it whenever the element changes.
  _onTimelineMountWithRef = (ref: HTMLElement | null) => {
    if (this._removeWheelListener) {
      this._removeWheelListener();
      this._removeWheelListener = null;
    }

    if (!ref) {
      return;
    }

    // without pinning to a const variable, Flow isn't sure that we don't change
    // the `ref` variable in some of the function calls below, and therefore
    // that it won't be null.
    const existingRef = ref;

    // Disable pinch-to-zoom and ctrl + wheel otherwise, on the timeline.
    // Indeed the users are used to this gesture  to zoom in in our charts, and
    // may use the same gesture elsewhere because of their habits, however this
    // doesn't do what they expect and instead zooms in the page, which is distracting.
    existingRef.addEventListener('wheel', this.preventPinchToZoom, {
      passive: false,
    });

    this._removeWheelListener = () => {
      existingRef.removeEventListener('wheel', this.preventPinchToZoom, {
        passive: false,
      });
    };
  };

  componentWillUnmount() {
    if (this._removeWheelListener) {
      this._removeWheelListener();
      this._removeWheelListener = null;
    }
  }

  render() {
    return <FullTimeline innerElementRef={this._onTimelineMountWithRef} />;
  }
}
