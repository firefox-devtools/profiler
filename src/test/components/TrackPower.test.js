/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import type { IndexIntoSamplesTable, CssPixels } from 'firefox-profiler/types';

import * as React from 'react';
import { Provider } from 'react-redux';
import { fireEvent } from '@testing-library/react';

import { render, screen } from 'firefox-profiler/test/fixtures/testing-library';

import { updatePreviewSelection } from 'firefox-profiler/actions/profile-view';
import { TrackPower } from '../../components/timeline/TrackPower';
import { ensureExists } from '../../utils/flow';

import {
  autoMockCanvasContext,
  flushDrawLog,
} from '../fixtures/mocks/canvas-context';
import { mockRaf } from '../fixtures/mocks/request-animation-frame';
import { storeWithProfile } from '../fixtures/stores';
import {
  addRootOverlayElement,
  removeRootOverlayElement,
  getMouseEvent,
} from '../fixtures/utils';
import {
  getProfileFromTextSamples,
  getCounterForThreadWithSamples,
} from '../fixtures/profiles/processed-profile';
import { autoMockElementSize } from '../fixtures/mocks/element-size';
import { autoMockIntersectionObserver } from '../fixtures/mocks/intersection-observer';

// The following constants determine the size of the drawn graph.
const SAMPLE_COUNT = 12;
const PIXELS_PER_SAMPLE = 10;
const GRAPH_WIDTH = PIXELS_PER_SAMPLE * SAMPLE_COUNT;
const GRAPH_HEIGHT = 10;

function getSamplesPixelPosition(
  sampleIndex: IndexIntoSamplesTable,
  samplePosition
): CssPixels {
  // Compute the pixel position of the center of a given sample.
  return sampleIndex * PIXELS_PER_SAMPLE + PIXELS_PER_SAMPLE * samplePosition;
}

/**
 * This test verifies that the power track can draw a graph.
 */
describe('TrackPower', function () {
  function setup() {
    const { profile } = getProfileFromTextSamples(
      Array(SAMPLE_COUNT).fill('A').join('  ')
    );
    const threadIndex = 0;
    const thread = profile.threads[threadIndex];
    const sampleTimes = ensureExists(thread.samples.time);
    // Changing one of the sample times, so we can test different intervals.
    sampleTimes[1] = 1.5; // It was 1 before.
    // Ensure some samples are very close to each other, to exercise
    // the max min decimation algorithm.
    for (let i = 7; i < sampleTimes.length - 1; ++i) {
      sampleTimes[i] = 7 + i / 100;
    }
    profile.counters = [
      getCounterForThreadWithSamples(
        thread,
        threadIndex,
        {
          time: sampleTimes.slice(),
          // Power usage numbers. They are pWh so they are pretty big.
          count: [
            10000, 40000, 50000, 100000, 2000000, 5000000, 30000, 1000000,
            20000, 1, 12000, 100000,
          ],
          length: SAMPLE_COUNT,
        },
        'SystemPower',
        'power'
      ),
    ];
    const store = storeWithProfile(profile);
    const { getState, dispatch } = store;
    const flushRafCalls = mockRaf();

    const renderResult = render(
      <Provider store={store}>
        <TrackPower counterIndex={0} />
      </Provider>
    );
    const { container } = renderResult;

    // WithSize uses requestAnimationFrame
    flushRafCalls();

    const canvas = ensureExists(
      container.querySelector('.timelineTrackPowerCanvas'),
      `Couldn't find the power canvas, with selector .timelineTrackPowerCanvas`
    );
    const getTooltipContents = () =>
      document.querySelector('.timelineTrackPowerTooltip');
    const getPowerDot = () =>
      container.querySelector('.timelineTrackPowerGraphDot');
    const moveMouseAtCounter = (index, pos) =>
      fireEvent(
        canvas,
        getMouseEvent('mousemove', {
          pageX: getSamplesPixelPosition(index, pos),
        })
      );

    return {
      ...renderResult,
      dispatch,
      getState,
      profile,
      thread,
      store,
      threadIndex,
      canvas,
      getTooltipContents,
      moveMouseAtCounter,
      flushRafCalls,
      getPowerDot,
    };
  }

  autoMockCanvasContext();
  autoMockElementSize({ width: GRAPH_WIDTH, height: GRAPH_HEIGHT });
  autoMockIntersectionObserver();
  beforeEach(addRootOverlayElement);
  afterEach(removeRootOverlayElement);

  it('matches the component snapshot', () => {
    const { container } = setup();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches the 2d canvas draw snapshot', () => {
    const { flushRafCalls } = setup();
    flushRafCalls();
    expect(flushDrawLog()).toMatchSnapshot();
  });

  it('can create a tooltip', function () {
    const { moveMouseAtCounter, getTooltipContents, canvas } = setup();
    expect(getTooltipContents()).toBeFalsy();
    moveMouseAtCounter(1, 0.5);
    expect(getTooltipContents()).toBeTruthy();
    fireEvent.mouseLeave(canvas);
    expect(getTooltipContents()).toBeFalsy();
  });

  it('has a tooltip that matches the snapshot', function () {
    const { moveMouseAtCounter, getTooltipContents } = setup();
    // We are hovering exactly between 4th and 5th counter. That's why it should
    // show the 5th counter.
    // Here are the values we'll get in the tooltip:
    // 5th counter value is 5000000 pWh, that is 5 µWh. Over 1ms, this means
    // 18W (5 * 1000 * 3600 / 1000^2).
    // Over the full range, we get 7.240 µWh, therefore we'll see in the tooltip
    // 0.007 mWh.
    moveMouseAtCounter(4, 0.5);
    expect(getTooltipContents()).toMatchSnapshot();
  });

  it('draws a dot on the graph', function () {
    const { moveMouseAtCounter, getPowerDot } = setup();
    expect(getPowerDot()).toBeFalsy();
    moveMouseAtCounter(1, 0.5);
    expect(getPowerDot()).toBeTruthy();
  });

  it('can draw a dot on both extremes of the graph', function () {
    const { moveMouseAtCounter, getPowerDot } = setup();
    expect(getPowerDot()).toBeFalsy();
    moveMouseAtCounter(0, 0.25);
    expect(getPowerDot()).toBeTruthy();
    moveMouseAtCounter(7, 0);
    expect(getPowerDot()).toBeTruthy();
  });

  it('draws a dot that matches the snapshot', function () {
    const { moveMouseAtCounter, getPowerDot } = setup();
    moveMouseAtCounter(1, 0.5);
    expect(getPowerDot()).toMatchSnapshot();
  });

  it('shows a tooltip with a Power: line and a power unit', function () {
    const { dispatch, moveMouseAtCounter } = setup();
    dispatch(
      updatePreviewSelection({
        hasSelection: true,
        isModifying: false,
        selectionStart: 5,
        selectionEnd: 6,
      })
    );
    moveMouseAtCounter(3, 0);
    // 100000pWh spent over 1ms is 360mW
    // Note: Fluent adds isolation characters \u2068 and \u2069 around variables.
    expect(screen.getByText(/Power:/).nextSibling).toHaveTextContent(
      '360\u2069 mW'
    );
    // Over the full range, we get 8.352 µWh, therefore we'll see in the tooltip
    // 8.4 µWh.
    expect(screen.getByText(/visible range:/).nextSibling).toHaveTextContent(
      '8.4\u2069 µWh'
    );
    // Over the preview selection, we get 5 µWh which shows up as 5.0 µWh.
    expect(
      screen.getByText(/Energy used in the current selection:/).nextSibling
    ).toHaveTextContent('5.0\u2069 µWh');
    expect(
      screen.getByText(/Average power in the current selection/).nextSibling
    ).toHaveTextContent('18.0\u2069 W');
  });

  it('does not break when the selection is empty', function () {
    const { dispatch, moveMouseAtCounter } = setup();
    dispatch(
      updatePreviewSelection({
        hasSelection: true,
        isModifying: false,
        selectionStart: 6,
        selectionEnd: 6,
      })
    );
    moveMouseAtCounter(3, 0);
    // 100000pWh spent over 1ms is 360mW
    // Note: Fluent adds isolation characters \u2068 and \u2069 around variables.
    expect(screen.getByText(/Power:/).nextSibling).toHaveTextContent(
      '360\u2069 mW'
    );
    // Over the full range, we get 8.352 µWh, therefore we'll see in the tooltip
    // 8.4 µWh.
    expect(screen.getByText(/visible range:/).nextSibling).toHaveTextContent(
      '8.4\u2069 µWh'
    );

    // The preview selection being empty, these 2 lines are not useful and
    // therefore not rendered.
    expect(
      screen.queryByText(/Energy used in the current selection:/)
    ).not.toBeInTheDocument();
    expect(
      screen.queryByText(/Average power in the current selection/)
    ).not.toBeInTheDocument();
  });
});
