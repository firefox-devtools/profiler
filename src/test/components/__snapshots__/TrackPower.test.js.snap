// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TrackPower draws a dot that matches the snapshot 1`] = `
<div
  class="timelineTrackPowerGraphDot"
  style="left: 15px; top: 24.872px; background-color: rgb(74, 74, 79);"
/>
`;

exports[`TrackPower has a tooltip that matches the snapshot 1`] = `
<div
  class="timelineTrackPowerTooltip"
>
  <div
    class="tooltipDetails"
  >
    <div
      class="tooltipLabel"
    >
      Power
      :
    </div>
    ⁨18.0⁩ W
    <div
      class="tooltipLabel"
    >
      Energy used in the visible range
      :
    </div>
    ⁨8.4⁩ µWh (⁨0.004⁩ mg CO₂e)
  </div>
</div>
`;

exports[`TrackPower matches the 2d canvas draw snapshot 1`] = `
Array [
  Array [
    "clearRect",
    0,
    0,
    120,
    25,
  ],
  Array [
    "set lineWidth",
    2,
  ],
  Array [
    "set lineJoin",
    "bevel",
  ],
  Array [
    "set strokeStyle",
    "#737373",
  ],
  Array [
    "set fillStyle",
    "#73737388",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "moveTo",
    0,
    26,
  ],
  Array [
    "lineTo",
    15,
    24,
  ],
  Array [
    "lineTo",
    20,
    24,
  ],
  Array [
    "lineTo",
    30,
    24,
  ],
  Array [
    "lineTo",
    40,
    15,
  ],
  Array [
    "lineTo",
    50,
    1,
  ],
  Array [
    "lineTo",
    60,
    24,
  ],
  Array [
    "lineTo",
    71,
    20,
  ],
  Array [
    "lineTo",
    71,
    24,
  ],
  Array [
    "lineTo",
    71,
    15,
  ],
  Array [
    "lineTo",
    71,
    18,
  ],
  Array [
    "lineTo",
    110,
    24,
  ],
  Array [
    "lineTo",
    120,
    24,
  ],
  Array [
    "stroke",
  ],
  Array [
    "lineTo",
    120,
    25,
  ],
  Array [
    "lineTo",
    0,
    25,
  ],
  Array [
    "fill",
  ],
]
`;

exports[`TrackPower matches the component snapshot 1`] = `
<div
  class="timelineTrackPower"
  style="height: 25px; --graph-height: 25px;"
>
  <div
    class="timelineTrackPowerGraph"
  >
    <div>
      <canvas
        class="timelineTrackPowerCanvas"
        height="25"
        width="120"
      />
    </div>
    <div
      class="timelineEmptyThreadIndicator"
    />
  </div>
</div>
`;
